# Лабораторная работа №4 Генерация текстов на основе LSTM

## Импорт библиотек

In [6]:
import re

import numpy as np
import pandas as pd


import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize

import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torchtext.vocab import build_vocab_from_iterator

from sklearn.model_selection import train_test_split

import sqlite3
import zipfile

from tqdm.auto import tqdm

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/kirillanpilov/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


## Device

In [7]:
device = torch.device("cpu")

## Функция обучения

In [27]:
def train(model: nn.Module, crterion, optimizer , n_epochs, train_loader, test_loader):
    loss_train = []
    accuracy_train = []

    for epoch in range(n_epochs):
        model.train()
        for batch in tqdm(train_loader , desc=f'Training epoch {epoch + 1}:'):
            inputs = batch['input']
            labels = batch['label']
            output = model(inputs)
            loss = criterion(output, labels)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()


        model.eval()
        total = 0
        correct = 0
        for batch in tqdm(test_loader , desc=f'Testing epoch {epoch + 1}:'):
              inputs = batch['input']
              labels = batch['label']
              output = model(inputs)
              _, predicted = torch.max(output.data, 1)
              labels = torch.nonzero(labels == 1)[:,1]

              total += labels.size(0)
              correct += (predicted == labels).sum().item()

        test_accuracy = correct / total
        accuracy_train.append(test_accuracy)

        print('Epoch [{}/{}], Loss: {:.4f}, Test Accuracy: {:.2f}%'.format(epoch + 1, n_epochs, loss.item(), test_accuracy * 100))
        loss_train.append(loss.item())
        generate_text(model, "я люблю ", num = 100)


## Задание 1. Загрузите текст из произведений Ницше
('nietzsche.txt', origin='https://s3.amazonaws.com/text-datasets/nietzsche.txt').


Выведете следующее:
- А) длину всего корпуса;
- Б) количество предложений;
- В) сколько всего символов используется?


In [8]:
with open('nietzsche.txt', 'r') as file:
    lines = file.read()

text_nietzsche = lines
text_nietzsche[:100]

'PREFACE\n\n\nSUPPOSING that Truth is a woman--what then? Is there not ground\nfor suspecting that all ph'

In [9]:
text_nietzsche = text_nietzsche.lower()
text_nietzsche = re.sub(r'\s', ' ', text_nietzsche) # replace \n \t and etc.
text_nietzsche = re.sub(r'\s{2,}', ' ', text_nietzsche) # replace repeated whitespace

### Длина всего корпуса

In [10]:
len(text_nietzsche)

598861

### Количество предложений

In [11]:
len(sent_tokenize(text_nietzsche))

2864

### Cколько всего символов используется?

In [12]:
len(set(text_nietzsche))

56

## Задание 2. Сократите текст наполовину избыточными последовательностями символов maxlen

In [13]:
vocab = build_vocab_from_iterator(text_nietzsche, specials=["<unk>"])
token2inx = vocab.get_stoi()
inx2token = {inx: token for token, inx in vocab.get_stoi().items()}

In [14]:
token2inx

{'é': 55,
 'æ': 54,
 '[': 49,
 '7': 48,
 'ä': 53,
 ']': 50,
 '8': 46,
 '4': 44,
 '2': 41,
 'z': 40,
 "'": 39,
 '9': 51,
 ')': 36,
 '(': 35,
 '?': 34,
 'q': 32,
 'j': 31,
 ';': 30,
 '1': 37,
 ':': 29,
 '6': 47,
 '3': 43,
 'r': 9,
 's': 8,
 'd': 12,
 'ë': 56,
 '=': 38,
 'k': 26,
 'n': 7,
 'h': 10,
 'u': 14,
 '0': 52,
 'p': 17,
 ',': 21,
 '"': 27,
 't': 3,
 'i': 4,
 '-': 24,
 '5': 45,
 'a': 5,
 'e': 2,
 'v': 23,
 'o': 6,
 '<unk>': 0,
 '.': 25,
 'c': 13,
 'f': 15,
 'm': 16,
 '!': 33,
 'l': 11,
 'g': 18,
 '_': 42,
 'y': 19,
 'w': 20,
 'b': 22,
 ' ': 1,
 'x': 28}

In [15]:
maxlen = 40
step = 3
sentences = []
next_chars = []
for i in range(0, len(text_nietzsche) - maxlen, step):
    sentences.append(text_nietzsche[i: i + maxlen])
    next_chars.append(text_nietzsche[i + maxlen])
print('nb sequences:', len(sentences))

print('Vectorization...')
x = np.zeros((len(sentences), maxlen, len(token2inx)))
y = np.zeros((len(sentences), len(token2inx)))
for i, sentence in enumerate(sentences):
    for t, char in enumerate(sentence):
        x[i, t, token2inx[char]] = 1
    y[i, token2inx[next_chars[i]]] = 1

nb sequences: 199607
Vectorization...


In [16]:
x.shape

(199607, 40, 57)

In [17]:
sentences[0]

'preface supposing that truth is a woman-'

In [18]:
x[0][0]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0.])

У нас в датасете 199607 40 символьных отрезков и ответ следущая буква. В каждом sample 40 списков one-hot, где одна единица обозначает букву.

## Задание 3. Создайте модель LSTM для генерации текста.

- А) Напишите вспомогательную функцию для выборки индекса из массива вероятностей
- Б) Напишите функцию, которая будет вызываться в конце каждой эпохи и печатать сгенерированный текст
- В) Запустите модель на обучение Имейте ввиду, что требуется не менее 20 эпох, прежде чем сгенерированный текст начнет звучать связно. Рекомендуется запускать этот скрипт на графическом процессоре, так как рекуррентные сети требуют довольно больших вычислительных затрат.
- Г) Проверьте работу модели в онлайн режиме.

### Train and test

In [19]:
X_train, X_test , y_train, y_test = train_test_split(x,y, test_size = 0.2)

In [20]:
len(X_train) , len(X_test) , len(y_train) , len(y_test)

(159685, 39922, 159685, 39922)

### Dataset

In [18]:
class CustomDataset(Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        input = torch.Tensor(self.data[idx]).to(device)
        label = torch.Tensor(self.labels[idx]).to(device)
        sample = {'input': input, 'label': label}
        return sample

In [22]:
train_dataset= CustomDataset(X_train, y_train)

test_dataset= CustomDataset(X_test, y_test)

### DataLoader

In [23]:
train_dataloader = DataLoader(train_dataset,  batch_size=256)

test_dataloader = DataLoader(test_dataset, batch_size=256)

### Функция для генерации текста

In [19]:
def generate_text(model, text, num = 100):
      model.eval()
      for i in range(num):
        x = torch.zeros(len(text), vocab_size).to(device)
        for j, char in enumerate(text):
          x[j, token2inx[char]] = 1
        x = x.unsqueeze(0)


        output = model(x)
        _, predicted = torch.max(output.data, 1)

        text += inx2token[int(predicted)]
      print(text)


### Модель

In [20]:
class AnpilovGpt(nn.Module):
    def __init__(self, vocab_size, hidden_size1, hidden_size2 ):
        super().__init__()
        self.lstm = nn.LSTM(vocab_size, hidden_size1,  batch_first=True) # (batch_size, seq_len, input_size)
        self.linear = nn.Linear(in_features = hidden_size1 , out_features = hidden_size2 )
        self.projection = nn.Linear(in_features = hidden_size2 , out_features = vocab_size) # vocab_size = num_classes

        self.dropout = nn.Dropout(p = 0.25) # defult p = 0.5
        self.tanh = nn.Tanh() # against vanish gradient

    def forward(self, input):
        output, _ = self.lstm(input)  # [batch_size, seq_len, hidden_dim]
        output = output[:, -1]
        output = self.dropout(self.linear(self.tanh(output)))  # [batch_size, seq_len, hidden_dim]
        projection = self.projection(self.tanh(output))
        return projection


### Проверка модели на свой текст

### Обучение

In [26]:
n_epochs = 20
vocab_size = len(vocab) # 57
hidden_size1 = 256
hidden_size2 = 128


model = AnpilovGpt(vocab_size, hidden_size1, hidden_size2).to(device)
criterion = nn.CrossEntropyLoss().to(device)
optimizer = torch.optim.Adam(model.parameters()) # lr = 0.001


train(model, criterion, optimizer, n_epochs, train_dataloader, test_dataloader)

Training epoch 1::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 1::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [1/20], Loss: 2.3763, Test Accuracy: 32.18%
i love the the the the the the the the the the the the the the the the the the the the the the the the the 


Training epoch 2::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 2::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [2/20], Loss: 2.1468, Test Accuracy: 37.85%
i love the the the the the the the the the the the the the the the the the the the the the the the the the 


Training epoch 3::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 3::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [3/20], Loss: 1.9647, Test Accuracy: 42.23%
i love the sould the sould the sould the sould the sould the sould the sould the sould the sould the sould 


Training epoch 4::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 4::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [4/20], Loss: 1.7390, Test Accuracy: 45.67%
i love the some the some the some the some the some the some the some the some the some the some the some t


Training epoch 5::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 5::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [5/20], Loss: 1.6531, Test Accuracy: 49.07%
i love the somether the somether the somether the somether the somether the somether the somether the somet


Training epoch 6::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 6::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [6/20], Loss: 1.5642, Test Accuracy: 50.98%
i love the self-consines the self-consines the self-consines the self-consines the self-consines the self-c


Training epoch 7::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 7::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [7/20], Loss: 1.4462, Test Accuracy: 52.36%
i love to the soul of the self-spirits of the self-spirits of the self-spirits of the self-spirits of the s


Training epoch 8::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 8::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [8/20], Loss: 1.4198, Test Accuracy: 53.26%
i love of the self-spirits of the self-spirits of the self-spirits of the self-spirits of the self-spirits 


Training epoch 9::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 9::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [9/20], Loss: 1.3153, Test Accuracy: 53.97%
i love of the self-spirits of the sense of the sense of the sense of the sense of the sense of the sense of


Training epoch 10::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 10::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [10/20], Loss: 1.2941, Test Accuracy: 54.67%
i love of the senses of strength and strong and the some can be no longer the complete the considerate stan


Training epoch 11::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 11::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [11/20], Loss: 1.2398, Test Accuracy: 54.74%
i love of the self-strungly and so that the senses of strength and supposing the compreter the senses of st


Training epoch 12::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 12::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [12/20], Loss: 1.1904, Test Accuracy: 54.85%
i love of the senses of supposing the considerate strength and strength and strength and strength and stren


Training epoch 13::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 13::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [13/20], Loss: 1.1372, Test Accuracy: 54.69%
i love of the feelings of strength and strength and strength and strength and strength and strength and str


Training epoch 14::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 14::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [14/20], Loss: 1.1213, Test Accuracy: 54.27%
i love out of the senses the powerly deligion and every one stands to super the constitus of the senses the


Training epoch 15::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 15::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [15/20], Loss: 1.0756, Test Accuracy: 54.17%
i love out of the words and all that is the preserved the existence of the powerful still rearer to the poi


Training epoch 16::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 16::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [16/20], Loss: 0.9983, Test Accuracy: 53.74%
i love out of the more spirits and so that the senses that the self-contradict that it is anything profound


Training epoch 17::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 17::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [17/20], Loss: 0.9334, Test Accuracy: 53.20%
i love of the point the obligation of the powerly called or the company rage of the world of a termisting o


Training epoch 18::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 18::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [18/20], Loss: 0.9114, Test Accuracy: 52.73%
i love of fact, the self-sensually in the most diving constituent of mankind that they are good nature, the


Training epoch 19::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 19::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [19/20], Loss: 0.8525, Test Accuracy: 52.40%
i love of fact that every stronger as a called exere such a tendous of something of mankind is a sort of su


Training epoch 20::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 20::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [20/20], Loss: 0.8334, Test Accuracy: 52.09%
i love of for his philosophers have look work soul what in the power of spirits and sometom and others in a


### Обучение с dropout = 0.25

In [27]:
n_epochs = 20
vocab_size = len(vocab) # 57
hidden_size1 = 256
hidden_size2 = 128


model = AnpilovGpt(vocab_size, hidden_size1, hidden_size2).to(device)
criterion = nn.CrossEntropyLoss().to(device)
optimizer = torch.optim.Adam(model.parameters()) # lr = 0.001


train(model, criterion, optimizer, n_epochs, train_dataloader, test_dataloader)

Training epoch 1::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 1::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [1/20], Loss: 2.4513, Test Accuracy: 31.90%
i love the the the the the the the the the the the the the the the the the the the the the the the the the 


Training epoch 2::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 2::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [2/20], Loss: 2.2280, Test Accuracy: 37.53%
i love the the the the the the the the the the the the the the the the the the the the the the the the the 


Training epoch 3::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 3::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [3/20], Loss: 2.0441, Test Accuracy: 41.09%
i love the and and and and and and and and and and and and and and and and and and and and and and and and 


Training epoch 4::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 4::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [4/20], Loss: 1.8719, Test Accuracy: 44.17%
i love the some the and and and and and and and and and and and and and and and and and and and and and and


Training epoch 5::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 5::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [5/20], Loss: 1.7453, Test Accuracy: 47.00%
i love the some the some the some the some the some the some the some the some the some the some the some t


Training epoch 6::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 6::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [6/20], Loss: 1.6472, Test Accuracy: 49.50%
i love and sermand and still the some the some the some the some the some the some the some the some the so


Training epoch 7::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 7::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [7/20], Loss: 1.5262, Test Accuracy: 51.08%
i love and sermand and an an are the something the something the something the something the something the 


Training epoch 8::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 8::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [8/20], Loss: 1.4503, Test Accuracy: 52.45%
i love and sermand and string the something the something the something the something the something the som


Training epoch 9::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 9::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [9/20], Loss: 1.3855, Test Accuracy: 53.27%
i love and suppose the self-constions of the something the something the something the something the someth


Training epoch 10::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 10::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [10/20], Loss: 1.3489, Test Accuracy: 54.00%
i love and suppose the love of the something the something the something the something the something the so


Training epoch 11::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 11::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [11/20], Loss: 1.2724, Test Accuracy: 54.51%
i love and suppose the lover of the something the constinate the existence of the something the constinate 


Training epoch 12::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 12::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [12/20], Loss: 1.2306, Test Accuracy: 54.85%
i love and strange of the stands of the strang himself of the something the constitute the serman so the ex


Training epoch 13::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 13::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [13/20], Loss: 1.2477, Test Accuracy: 54.99%
i love and strange of the stands of the strang himself of the something the constinut of the something the 


Training epoch 14::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 14::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [14/20], Loss: 1.2204, Test Accuracy: 54.74%
i love and simplication of the consequently the consequently the consequently the consequently the conseque


Training epoch 15::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 15::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [15/20], Loss: 1.1686, Test Accuracy: 54.50%
i love and simple the possessition of the consequences of such a man of the consequences of such a man of t


Training epoch 16::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 16::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [16/20], Loss: 1.1333, Test Accuracy: 54.41%
i love and strangly as a part of the common as a philosophy of the are so the ever reade the ever been more


Training epoch 17::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 17::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [17/20], Loss: 1.0663, Test Accuracy: 53.94%
i love of the commander the consciences is so far as a philosopher has a stronger of mankind the conscience


Training epoch 18::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 18::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [18/20], Loss: 1.0560, Test Accuracy: 53.76%
i love is not a feal of the artists of the postions of the soul of the stands of the stands of the stands o


Training epoch 19::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 19::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [19/20], Loss: 1.0076, Test Accuracy: 53.12%
i love of the conduct the deerers of the artists of his properiting the postions of the souls of the genera


Training epoch 20::   0%|          | 0/624 [00:00<?, ?it/s]

Testing epoch 20::   0%|          | 0/156 [00:00<?, ?it/s]

Epoch [20/20], Loss: 0.9581, Test Accuracy: 52.50%
i love of the conduct therefore, that is to say that the sense of the consciences is so far allo of the age


### Проверка модели в онлайн режиме.

In [28]:
text = "what a day"
generate_text(model, text, 1000)

what a day been position of the conduct therefore, that is to say that the sense of the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the agesome the consciences is so far allo of the ages

## Задание 4. Создайте самостоятельно генерацию текста для РУССКОЯЗЫЧНОГО НАБОРА глав Wikibooks.
Полный текст Wikibooks содержит более 270000 глав на 12 языках https://www.kaggle.com/datasets/dhruvildave/wikibooks-dataset/data


In [3]:
from google.colab import drive
drive.mount('/content/drive')

ModuleNotFoundError: No module named 'google'

In [8]:
zf = zipfile.ZipFile('archive.zip')
for file in tqdm(zf.infolist()):
    zf.extract(file)

  0%|          | 0/1 [00:00<?, ?it/s]

In [29]:
conn = sqlite3.connect('wikibooks.sqlite')

cursor = conn.cursor()

cursor.execute("SELECT * from ru")

raw_eng_text = cursor.fetchall()

In [30]:
columns = [i[0] for i in cursor.description]
data = pd.DataFrame(raw_eng_text, columns=columns)

cursor.close()
conn.close()
text = " ".join(data['body_text'])

In [31]:
text = text.lower()
text = re.sub(r'[^а-яА-ЯёЁ ]', '', text)
text = re.sub(r'\s', ' ', text) # replace \n \t and etc.
text = re.sub(r'\s{2,}', ' ', text) # replace repeated whitespace
text[:100]

'рабочая станциясерверперсональный компьютер в википедии имеется статья по теме свидетельство частног'

In [32]:
len(text)

38837958

In [33]:
vocab = build_vocab_from_iterator(text, specials=["<unk>"])
token2inx = vocab.get_stoi()
inx2token = {inx: token for token, inx in vocab.get_stoi().items()}

In [34]:
token2inx

{'щ': 31,
 'ш': 30,
 'ф': 29,
 'э': 32,
 'ю': 28,
 'ё': 33,
 'ж': 26,
 'х': 25,
 'й': 24,
 'г': 22,
 'я': 17,
 'д': 15,
 'б': 21,
 'п': 14,
 'м': 13,
 'ъ': 34,
 ' ': 1,
 'к': 12,
 'з': 19,
 'л': 11,
 'ч': 23,
 'и': 4,
 'в': 10,
 'т': 7,
 'ь': 20,
 'у': 16,
 'н': 6,
 'а': 5,
 'р': 8,
 'о': 2,
 'ц': 27,
 'е': 3,
 'ы': 18,
 'с': 9,
 '<unk>': 0}

In [35]:
maxlen = 40
step = 3
sentences = []
next_chars = []
for i in range(0, len(text) - maxlen, step):
    sentences.append(text[i: i + maxlen])
    next_chars.append(text[i + maxlen])
    if len(sentences) > 1_000_000:
      break
print('nb sequences:', len(sentences))

print('Vectorization...')
x = np.zeros((len(sentences), maxlen, len(token2inx)))
y = np.zeros((len(sentences), len(token2inx)))
for i, sentence in enumerate(sentences):
    for t, char in enumerate(sentence):
        x[i, t, token2inx[char]] = 1
    y[i, token2inx[next_chars[i]]] = 1

nb sequences: 1000001
Vectorization...


In [36]:
X_train, X_test , y_train, y_test = train_test_split(x,y, test_size = 0.2)

In [37]:
train_dataset = CustomDataset(X_train, y_train)

test_dataset = CustomDataset(X_test, y_test)

In [38]:
train_dataloader = DataLoader(train_dataset,  batch_size=256)

test_dataloader = DataLoader(test_dataset, batch_size=256)

In [39]:
n_epochs = 20
vocab_size = len(vocab) # 57
hidden_size1 = 256
hidden_size2 = 128


model_wiki = AnpilovGpt(vocab_size, hidden_size1, hidden_size2).to(device)
criterion = nn.CrossEntropyLoss().to(device)
optimizer = torch.optim.Adam(model_wiki.parameters()) # lr = 0.001


train(model_wiki, criterion, optimizer, n_epochs, train_dataloader, test_dataloader)

Training epoch 1::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 1::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [1/20], Loss: 1.8251, Test Accuracy: 45.19%
я люблю по совтретельности по совтретельности по совтретельности по совтретельности по совтретельности по со


Training epoch 2::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 2::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [2/20], Loss: 1.5814, Test Accuracy: 50.87%
я люблю с помощью по создание при после постоянность по суммы по создание при после постоянность по суммы по


Training epoch 3::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 3::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [3/20], Loss: 1.4672, Test Accuracy: 53.29%
я люблю обостренного примера при помощи по сказанием с помощью подсидить в течение проведения и подсидить в 


Training epoch 4::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 4::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [4/20], Loss: 1.4719, Test Accuracy: 54.56%
я люблю и доступность по создание состояние по создании составляющие состояние по создании составляющие сост


Training epoch 5::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 5::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [5/20], Loss: 1.4231, Test Accuracy: 55.34%
я люблю оборудование с помощью составляющие составляющие подразделение просто должна быть в статистической с


Training epoch 6::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 6::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [6/20], Loss: 1.3833, Test Accuracy: 55.81%
я люблю получился в статистических предприятия при проведении создания по создания программы в течение прост


Training epoch 7::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 7::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [7/20], Loss: 1.3500, Test Accuracy: 56.12%
я люблю и в статистических процессов в состав между которые можно подвержать состав по создание составляют с


Training epoch 8::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 8::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [8/20], Loss: 1.3374, Test Accuracy: 56.36%
я люблю получаем в строке при помощи метод после объекта в статистических предприятия при помощи метод после


Training epoch 9::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 9::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [9/20], Loss: 1.3161, Test Accuracy: 56.52%
я люблю получаем в своей обсуждает в состав к которому создание проведения измерений в стаким образованиям с


Training epoch 10::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 10::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [10/20], Loss: 1.3101, Test Accuracy: 56.57%
я люблю получаем контроля в документе с помощью современных в строке по создании создания процесса и подстав


Training epoch 11::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 11::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [11/20], Loss: 1.2670, Test Accuracy: 56.61%
я люблю получение в статистических обслуживаются в состав в строке при проведении создания при проведении со


Training epoch 12::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 12::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [12/20], Loss: 1.2561, Test Accuracy: 56.62%
я люблю получилось на рабочих с помощью создания при помощи метод полученных по создании создания при помощи


Training epoch 13::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 13::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [13/20], Loss: 1.2346, Test Accuracy: 56.60%
я люблю получилось на разработан при помощи при помощи проведения графика в состав в зависимости от стреенир


Training epoch 14::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 14::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [14/20], Loss: 1.2183, Test Accuracy: 56.59%
я люблю производственные и соответствующих неверното при помощи при помощи стандартивального распределения п


Training epoch 15::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 15::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [15/20], Loss: 1.1806, Test Accuracy: 56.58%
я люблю в конструктура обеспечивать в составе консультанта на расходование в стандарте при отравлении которы


Training epoch 16::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 16::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [16/20], Loss: 1.2104, Test Accuracy: 56.47%
я люблю получило для подразделения и подарить получить создание конструкция то время то что в то время котор


Training epoch 17::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 17::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [17/20], Loss: 1.2214, Test Accuracy: 56.39%
я люблю и в структуре предприятия на расходование в стаканической объектом в статистических объектом с помощ


Training epoch 18::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 18::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [18/20], Loss: 1.2097, Test Accuracy: 56.39%
я люблю и то в консульсами по создавать с помощью подстроки на рабочих с тем что создать следующие положение


Training epoch 19::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 19::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [19/20], Loss: 1.2240, Test Accuracy: 56.38%
я люблю и то в консульсами на стандартной проблемам на сотрудникам должно быть создавать в составе одной и с


Training epoch 20::   0%|          | 0/3125 [00:00<?, ?it/s]

Testing epoch 20::   0%|          | 0/782 [00:00<?, ?it/s]

Epoch [20/20], Loss: 1.1847, Test Accuracy: 56.31%
я люблю и то в составе от которых выделяюте в структуре предприятия в течение периода которые можно получить


In [43]:
text = "физика это "
generate_text(model_wiki, text, 1000)

физика это может быть на разностью момент но в разделе проведения измерений в процессе с помощью применения конструкций для проведения измерений в течение части продукт по проверка на выражению с помощью проведения измерений в течение части продаж по возможности поставщику на разрешении с помощью применения конструкций для проведения измерений в течение части продукт по проверка на выражению с помощью проведения измерений в течение части продаж по возможности поставщику на разрешении с помощью применения конструкций для проведения измерений в течение части продукт по проверка на выражению с помощью проведения измерений в течение части продаж по возможности поставщику на разрешении с помощью применения конструкций для проведения измерений в течение части продукт по проверка на выражению с помощью проведения измерений в течение части продаж по возможности поставщику на разрешении с помощью применения конструкций для проведения измерений в течение части продукт по проверка на выражению с 